cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala

```diff
@@ -442,9 +448,67 @@ class BreadthFirstTraverser[+V <: Node](val graph: Graph[V], val dir: GraphDir,
     super.visitNode(node)
   }
 
-  def shouldBeEnqueued(color: Walk.NodeColor.Color): Boolean = {
-    color == Walk.NodeColor.Unenqueued
+  def shouldBeEnqueued(color: Color): Boolean = color == Walk.NodeColor.Unenqueued
+}
+
+/**
+ * Traverses in BFS order and keeps track of all shortest paths from any node in `homeNodeIds` to
+ * all target paths.  The traversal happens along all outbound nodes, it is suggested that `graph`
+ * storage direction is `GraphDir.OutDir`
+ * @param graph The graph to traverse on.  This graph need not necessarily be directed, but the
+ *              traversal will happen along all outbound nodes.
+ * @param homeNodeIds The sequence of source nodes to find all shortest paths.  Each shortest path must
+ *                    start with one of the home nodes.
+ * @param dir The direction of the graph traversal
+ * @param walkLimits  How the graph traversal will be limited
+ * @tparam V A subtype of `Node`
+ */
+class ShortestPathsTraverser[+V <: Node](val graph: Graph[V], val homeNodeIds: Seq[Int],
+                                        val dir: GraphDir = GraphDir.OutDir,
+                                        walkLimits: Walk.Limits = Walk.Limits())
+  extends DepthTracker with BoundedIterator[V] {
+
+  // Queueing parameters
+  override def nodePriority = GraphTraverserNodePriority.FIFO
+  override lazy val limits = walkLimits
+
+  // Accounting parameters
+  val numPaths = mutable.Map.empty[Int, Int]
+  val paths    = mutable.Map.empty[Int, Seq[Seq[Int]]]
+  homeNodeIds foreach { h => paths(h) = Seq(Seq()) }
+
+  /**
+   * To find all shortest paths, the chosen enqueued nodes must be selected carefully.
+   * In the filtering process, the shortest paths will be updated simultaneously.
+   * @return The next set of nodes to be enqueued
+   */
+  override protected def chooseNodes(node: Node): Seq[Int] = {
+    numPaths(node.id) = if (numPaths.contains(node.id)) numPaths(node.id) + 1 else 1
+
+    val currDepth = depth(node.id).get
+    val nextDepth = currDepth + 1
+
+    val index = numPaths(node.id) - 1
```

Comment:
looks like you are only accessing the tail of paths(node.id) using 'index', and if so no need to keep numPaths(node.id) at all, yes?